src/pip/_internal/index.py

```diff
@@ -764,7 +787,9 @@ def find_requirement(self, req, upgrade):
         Raises DistributionNotFound or BestVersionAlreadyInstalled otherwise
         """
         candidates = self.find_candidates(req.name, req.specifier)
-        best_candidate = candidates.get_best()
+        # Get any hashes supplied by the user to filter candidates.
+        hashes = req.hashes(trust_internet=False)
```

Comment:
Conceptually, there should be a difference between supplying `--require-hashes` (i.e. hash mode with an empty hash list) and not (not hash mode). `candidates.get_best()` should only receieve the hash comparer in hash mode, and `None` otherwise (and `get_best` probably needs to check for `None` instead of a falsy hash comparer).

Reply:
I think `get_best` should not filter on falsy `hashes`. If I supply `--require-hashes` without any hashes and filter on `hashes`, this filters out all candidates and results in the following error:
```
ERROR: Could not find a version that satisfies the requirement <package_name> (from versions: ...)
```

If I don't filter on falsy `hashes` then using `--require-hashes` without any hashes results in:
```
ERROR: In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:
    tox from https://files.pythonhosted.org/packages/13/94/df88fb647f5f3d7c5afdb6c05bb56417c8bbbd0db0202f71448a040bba95/tox-3.9.0-py2.py3-none-any.whl#sha256=1b166b93d2ce66bb7b253ba944d2be89e0c9d432d49eeb9da2988b4902a4684e
```

I think the latter is much more helpful, and clearly some work went into it via the `MissingHashes` class.

Reply:
Although perhaps this is a better place to implement `MissingHashes` logic?

Reply:
Tried replacing `MissingHashes` with a `raise` in `get_best`. It works, except that it makes the error about missing hashes take precedence over the error about not having versions pinned. Currently not having versions pinned in hash mode takes precedence. I am not sure whether this order of precedence is intentional.

Reply:
Maybe it’s a good idea to use `MissingHashes` here (but I don’t think you can just move it here?)

Reply:
I think I would prefer to not filter, with a comment about how if `hashes` is `None`, an error will be raised in `unpack_url` which explains what hash to pin to. If desired, in a future PR I could refactor away `MissingHashes` and raise the same error in `get_best` instead (which is what I meant by "perhaps this is a better place to implement `MissingHashes` logic"). But I'd rather not mix refactoring with feature work, and it would require some care to get the order of precedence for error messages right (there's a lot of error handling logic right now in `RequirementPreparer.prepare_linked_requirement` which would need to get moved around).

Reply:
Do you have a plan how the refactoring would be done? If so, maybe it would be better to refactor first, and implement the feature on top of that.

Reply:
It turns out that such a refactoring has a downside the current implementation of `MissingHashes` will provide the correct hash to the user even if there is no remote hash, whereas if I were to move the `raise` into `get_best` it would happen before the candidate is downloaded and so I could only provide the hash to the user if it was provided by the index. So I think it is not worth doing.

Reply:
Alright then, let’s work on what we have and improve (if possible) in the future.